codex/blockexchange/engine/engine.nim

```diff
@@ -358,42 +361,49 @@ proc dropPeer*(b: BlockExcEngine, peer: PeerID) =
 proc taskHandler*(b: BlockExcEngine, task: BlockExcPeerCtx) {.gcsafe, async.} =
   trace "Handling task for peer", peer = task.id
 
-  var wantsBlocks = newAsyncHeapQueue[Entry](queueType = QueueType.Max)
-  # get blocks and wants to send to the remote
-  for e in task.peerWants:
-    if e.wantType == WantType.wantBlock:
-      await wantsBlocks.push(e)
+  # PART 1: Send to the peer blocks he wants to get,
+  # if they present in our local store
 
   # TODO: There should be all sorts of accounting of
   # bytes sent/received here
+
+  var wantsBlocks = task.peerWants.filterIt(it.wantType == WantType.wantBlock)
+
   if wantsBlocks.len > 0:
+    wantsBlocks.sort(SortOrder.Descending)
+
     let blockFuts = await allFinished(wantsBlocks.mapIt(
         b.localStore.getBlock(it.cid)
     ))
 
+    # Extract succesfully received blocks
     let blocks = blockFuts
-      .filterIt((not it.failed) and it.read.isOk)
-      .mapIt(!it.read)
+      .filterIt(it.completed and it.read.isOk and it.read.get.isSome)
```

Comment:
Well, that's appropriate definitions from chronos:
```Nim
type
  FutureState* {.pure.} = enum
    Pending, Finished, Cancelled, Failed

proc finished*(future: FutureBase): bool {.inline.} =
  ## Determines whether ``future`` has completed, i.e. ``future`` state changed
  ## from state ``Pending`` to one of the states (``Finished``, ``Cancelled``,
  ## ``Failed``).
  result = (future.state != FutureState.Pending)

proc cancelled*(future: FutureBase): bool {.inline.} =
  ## Determines whether ``future`` has cancelled.
  (future.state == FutureState.Cancelled)

proc failed*(future: FutureBase): bool {.inline.} =
  ## Determines whether ``future`` completed with an error.
  (future.state == FutureState.Failed)

proc completed*(future: FutureBase): bool {.inline.} =
  ## Determines whether ``future`` completed without an error.
  (future.state == FutureState.Finished)

proc done*(future: FutureBase): bool {.inline.} =
  ## This is an alias for ``completed(future)`` procedure.
  completed(future)
```

As you can see `completed` means exactly that - it was finished and it finished successfully. OTOH, `not it.failed` is true for Cancelled future which can't be read. So I suppose that I also catched a bug here. It will be nice to have a test.

Reply:
Ah, nice find - I was wrong then :)